### PR TITLE
Add ability for direct 3PlayMedia transcripts usage

### PR DESCRIPTION
**Pull request by wowkalucky**
General implementation of direct (streaming) transcripts usage from 3PlayMedia service.

Till the PR VideoXBlock already has ability to use transcripts from 3PlayMedia service. But those '3playmedia' transcripts were used by downloading them into xblock as course assets.
So, if any changes was made on the 3PlayMedia service to mirror those changes on the xblock User should manually perform downloading each time.

The Client requested ability of VideoXBlock to use 3PlayMedia's transcripts directly (so, changes on 3PM service take effect immediately).

And PR adds described functionality.

### CODE REVIEW

File: video_xblock/backends/base.py [hunk 1]
```diff
@@ -132,8 +132,8 @@ def advanced_fields(self):
         """
         return [
             'start_time', 'end_time', 'handout', 'transcripts',
-            'threeplaymedia_file_id', 'threeplaymedia_apikey', 'download_transcript_allowed',
-            'default_transcripts', 'download_video_allowed', 'download_video_url'
+            'threeplaymedia_file_id', 'threeplaymedia_apikey', 'direct_enabled',
```
Comment (z4y4ts) [inline video_xblock/backends/base.py]: `direct_enabled` -> `threeplaymedia_streaming` or `threeplaymedia_direct_enabled`

File: video_xblock/mixins.py [hunk 2]
```diff
@@ -58,6 +56,31 @@ class TranscriptsMixin(XBlock):
     TranscriptsMixin class to encapsulate transcripts-related logic.
     """
 
+    THREE_PLAY_MEDIA_API_DOMAIN = 'https://static.3playmedia.com/'
+
+    direct_enabled = Boolean(
```
Comment (z4y4ts) [inline video_xblock/mixins.py]: `direct_enabled` -> `threeplaymedia_streaming` or `threeplaymedia_direct_enabled`

File: video_xblock/mixins.py [hunk 3]
```diff
@@ -58,6 +56,31 @@ class TranscriptsMixin(XBlock):
...
+        default=False,
+        display_name=_('Direct 3PlayMedia'),
+        scope=Scope.settings,
```
Comment (z4y4ts) [inline video_xblock/mixins.py]: Scope.content

File: video_xblock/mixins.py [hunk 4]
```diff
@@ -75,26 +98,34 @@ def convert_caps_to_vtt(caps):
...
             transcripts (unicode): Raw transcripts.
         """
-        transcripts = json.loads(transcripts) if transcripts else []
+        transcripts = self.get_enabled_transcripts()
         for tran in transcripts:
-            if not tran['url'].endswith('.vtt'):
+            if self.direct_enabled:
                 tran['url'] = self.runtime.handler_url(
-                    self, 'srt_to_vtt', query=tran['url']
+                    self, 'fetch_from_tree_play_media', query="{}={}".format(tran['lang_id'], tran['id'])
```
Comment (z4y4ts) [inline video_xblock/mixins.py]: tree -> three

File: video_xblock/mixins.py [hunk 5]
```diff
@@ -75,26 +98,34 @@ def convert_caps_to_vtt(caps):
...
                 )
+            else:
+                if not tran['url'].endswith('.vtt'):
```
Comment (z4y4ts) [inline video_xblock/mixins.py]: else if -> elif

File: video_xblock/mixins.py [hunk 6]
```diff
@@ -165,6 +196,81 @@ def convert_3playmedia_caps_to_vtt(self, caps, video_id, lang="en", lang_label="
...
+        :return: (generator of OrderedDicts) all transcript's data
+        """
+        results = self.get_available_3pm_transcripts(self.threeplaymedia_file_id, self.threeplaymedia_apikey)
+        errors = isinstance(results, dict) and results.get('errors')
+        log.debug("Fetched 3PM transcripts:\n{}".format(results))
+        if errors:
+            log.error("3PlayMedia transcripts fetching API request has failed!\n{}".format(
+                results.get('errors')
+            ))
+            raise StopIteration
```
Comment (z4y4ts) [inline video_xblock/mixins.py]: This one is pretty neat, actually.

File: video_xblock/mixins.py [hunk 7]
```diff
@@ -165,6 +196,81 @@ def convert_3playmedia_caps_to_vtt(self, caps, video_id, lang="en", lang_label="
...
+            domain=self.THREE_PLAY_MEDIA_API_DOMAIN,
+            file_id=self.threeplaymedia_file_id,
+            tid=tid,
+            api_key=self.threeplaymedia_apikey,
+            format_id=format_id
+        )
+        try:
+            content = requests.get(external_api_url).text
+        except Exception as exc:  # pylint: disable=broad-except
+            log.exception(exc.message)
```
Comment (z4y4ts) [inline video_xblock/mixins.py]: Exception info is always added to the logging message. So you don't need to use `exc.message`.
Instead, put here some meaningful message.

File: video_xblock/mixins.py [hunk 8]
```diff
@@ -285,6 +378,45 @@ def srt_to_vtt(self, request, _suffix=''):
         caps = requests.get(request.host_url + caps_path).text
         return Response(self.convert_caps_to_vtt(caps))
 
+    @XBlock.handler
+    def fetch_from_tree_play_media(self, request, _suffix=''):
```
Comment (z4y4ts) [inline video_xblock/mixins.py]: Typo - `tree`

File: video_xblock/mixins.py [hunk 9]
```diff
@@ -285,6 +378,45 @@ def srt_to_vtt(self, request, _suffix=''):
...
+            request (webob.Request):
+            suffix (string): not used
+        Returns:
+            webob.Response: (json) {'isValid': true/false}
+        """
+        api_key = request.json.get('api_key')
+        file_id = request.json.get('file_id')
+
+        if not (api_key and file_id):
+            return Response(status=400)
```
Comment (z4y4ts) [inline video_xblock/mixins.py]: Please provide a helpfull message, saying what's wrong with the supplied data.
E.g. `"api_key or file_id is missing"`

File: video_xblock/mixins.py [hunk 10]
```diff
@@ -285,6 +378,45 @@ def srt_to_vtt(self, request, _suffix=''):
...
+
+        results = self.get_available_3pm_transcripts(file_id, api_key)
+
+        is_valid = not isinstance(results, dict)
```
Comment (z4y4ts) [inline video_xblock/mixins.py]: This looks strange. Please reformulate in positive key or make it more specific.

File: video_xblock/static/js/studio-edit/studio-edit.js [hunk 11]
```diff
@@ -383,18 +383,75 @@ function StudioEditableXBlock(runtime, element) {
         });
     }
 
-    // Raccoongang changes
+    /**
+     * Validate if 3PlayMedia combination: fileId + apiKey is actual.
```
Comment (z4y4ts) [inline video_xblock/static/js/studio-edit/studio-edit.js]: Validate if 3PlayMedia options: fileId, apiKey.

File: video_xblock/static/js/studio-edit/studio-edit.js [hunk 12]
```diff
@@ -383,18 +383,75 @@ function StudioEditableXBlock(runtime, element) {
...
+     * @returns {boolean}
+     */
+    function validateThreePlayMediaConfig(data) {
+        var message;
+        var status;  // eslint-disable-line no-unused-vars
```
Comment (z4y4ts) [inline video_xblock/static/js/studio-edit/studio-edit.js]: Do you really need this var?

File: video_xblock/video_xblock.py [hunk 13]
```diff
@@ -782,3 +766,15 @@ def upload_default_transcript_handler(self, data, _suffix=''):
...
+        """
+        Get transcripts from different sources depending on current usage mode.
+        """
+        if self.direct_enabled:
+            transcripts = self.fetch_available_3pm_transcripts()
+        else:
+            transcripts = json.loads(self.transcripts) if self.transcripts else []
+
+        for transcript in transcripts:
+            yield transcript
```
Comment (z4y4ts) [inline video_xblock/video_xblock.py]: This for loop looks useless.
Why don't just `return transcripts`?

File: video_xblock/utils.py [hunk 14]
```diff
@@ -127,3 +128,8 @@ def normalize_transcripts(transcripts):
     for tr_dict in transcripts:
         tr_dict.setdefault('source', TranscriptSource.MANUAL)
     return transcripts
+
+
+Transcript = namedtuple('Transcript', [
+    'id', 'label', 'lang', 'lang_id', 'content', 'format', 'video_id', 'source', 'url'
```
Comment (z4y4ts) [inline video_xblock/utils.py]: Nice :)

File: video_xblock/mixins.py [hunk 15]
```diff
@@ -369,7 +501,7 @@ def player_state(self):
         """
         Return video player state as a dictionary.
         """
-        transcripts = json.loads(self.transcripts) if self.transcripts else []
+        transcripts = list(self.get_enabled_transcripts())
```
Comment (z4y4ts) [inline video_xblock/mixins.py]: Looking few lines below, this conversion to a list seems redundant.

File: video_xblock/static/js/studio-edit/transcripts-manual-upload.js [hunk 16]
```diff
@@ -53,12 +53,11 @@ function getTranscriptUrl(transcriptsArray, langCode) {
 /**
  * Validate transcript data before save it to video xblock.
  */
-function validateTranscripts(e, $langChoiceItem) {
+function validateTranscripts($langChoiceItem) {
     'use strict';
     var isValid = [];
     var $visibleLangChoiceItems = $langChoiceItem.find('li:visible');
     var urls;
-    e.preventDefault();
```
Comment (z4y4ts) [inline video_xblock/static/js/studio-edit/transcripts-manual-upload.js]: 👍

File: video_xblock/static/js/studio-edit/transcripts-manual-upload.js [hunk 17]
```diff
@@ -53,12 +53,11 @@ function getTranscriptUrl(transcriptsArray, langCode) {
 /**
  * Validate transcript data before save it to video xblock.
  */
-function validateTranscripts(e, $langChoiceItem) {
+function validateTranscripts($langChoiceItem) {
```
Comment (z4y4ts) [inline video_xblock/static/js/studio-edit/transcripts-manual-upload.js]: 👍

File: video_xblock/mixins.py (hunk 3 above)
Comment (wowkalucky) [inline video_xblock/mixins.py]: Ok. But what if we would like to use 3PM streaming only in one course Unit?

File: video_xblock/mixins.py [hunk 18]
```diff
@@ -165,6 +195,81 @@ def convert_3playmedia_caps_to_vtt(self, caps, video_id, lang="en", lang_label="
...
+            raise StopIteration
+        for transcript_data in results:
+            transcript = self.fetch_3pm_translation(transcript_data)
+            transcript_ordered_dict = transcript._asdict()
+            transcript_ordered_dict['content'] = ''  # we don't want to parse it to JSON
+            yield transcript_ordered_dict
+
+    def get_available_3pm_transcripts(self, file_id, apikey):
+        """
+        Make API request to fetch list of available transcripts for given file ID.
```
Comment (auraz) [inline video_xblock/mixins.py]: Add difference of return types of API in documentation

File: video_xblock/mixins.py [hunk 19]
```diff
@@ -165,6 +195,81 @@ def convert_3playmedia_caps_to_vtt(self, caps, video_id, lang="en", lang_label="
...
+        response = requests.get(
+            '{domain}files/{file_id}/transcripts?apikey={api_key}'.format(
+                domain=domain, file_id=file_id, api_key=apikey
+            )
+        )
+        if response.ok:
+            results = response.json()
+        else:
+            results = {"failed": True}
+        return results
```
Comment (auraz) [inline video_xblock/mixins.py]: Proposal: return response.json(), status

File: video_xblock/mixins.py [hunk 20]
```diff
@@ -165,6 +195,81 @@ def convert_3playmedia_caps_to_vtt(self, caps, video_id, lang="en", lang_label="
...
+        errors = isinstance(results, dict) and results.get('errors')
+        log.debug("Fetched 3PM transcripts:\n{}".format(results))
+        if errors:
+            log.error("3PlayMedia transcripts fetching API request has failed!\n{}".format(
+                results.get('errors')
+            ))
+            raise StopIteration
+        for transcript_data in results:
+            transcript = self.fetch_3pm_translation(transcript_data)
+            transcript_ordered_dict = transcript._asdict()
```
Comment (auraz) [inline video_xblock/mixins.py]: 1) Please add test for empty transcript
2) Think if we need to show this error to user.

File: video_xblock/mixins.py [hunk 21]
```diff
@@ -165,6 +195,81 @@ def convert_3playmedia_caps_to_vtt(self, caps, video_id, lang="en", lang_label="
...
+            transcript_ordered_dict['content'] = ''  # we don't want to parse it to JSON
```
Comment (auraz) [inline video_xblock/mixins.py]: extend comment.

File: video_xblock/mixins.py [hunk 22]
```diff
@@ -165,6 +195,81 @@ def convert_3playmedia_caps_to_vtt(self, caps, video_id, lang="en", lang_label="
...
+            file_id=self.threeplaymedia_file_id,
+            tid=tid,
+            api_key=self.threeplaymedia_apikey,
+            format_id=format_id
+        )
+        try:
+            content = requests.get(external_api_url).text
+        except Exception:  # pylint: disable=broad-except
+            log.exception(_("Transcript fetching failure: language [{}]").format(TPMApiLanguage(lang_id)))
+            return
```
Comment (auraz) [inline video_xblock/mixins.py]: Please think about consistency in corner cases. Here you do return, but in def __init__(self, language_id): you do raise.

File: video_xblock/mixins.py [hunk 23]
```diff
@@ -165,6 +195,81 @@ def convert_3playmedia_caps_to_vtt(self, caps, video_id, lang="en", lang_label="
...
+            lang=lang_code.iso_639_1_code,
+            lang_id=lang_id,
+            label=lang_label,
+            video_id=video_id,
+            format=format_id,
+            source=source,
+            url=external_api_url,
+        )
+
     def get_translations_from_3playmedia(self, file_id, apikey):
```
Comment (auraz) [inline video_xblock/mixins.py]: Plz rename into good naming.

File: video_xblock/mixins.py [hunk 24]
```diff
@@ -179,43 +284,30 @@ def get_translations_from_3playmedia(self, file_id, apikey):
             error_message (dict): Description of error.
```
Comment (auraz) [inline video_xblock/mixins.py]: Please unify docstrings

Comment (auraz) [inline video_xblock/mixins.py]: Plz clarify that file id is video id,  not transcript id.

File: video_xblock/mixins.py [hunk 25]
```diff
@@ -285,6 +377,45 @@ def srt_to_vtt(self, request, _suffix=''):
...
+        Proxy handler to hide real API url.
+
+        Arguments:
+            request (webob.Request): The request to handle
+            suffix (string): not used
+            query string: 'language_id=transcript_id'
+        Returns:
+            webob.Response: WebVTT transcripts wrapped in Response object.
+        """
+        lang_id, tid = request.query_string.split('=')
```
Comment (auraz) [inline video_xblock/mixins.py]: what is tid, please use better naming.

File: video_xblock/mixins.py [hunk 26]
```diff
@@ -285,6 +377,45 @@ def srt_to_vtt(self, request, _suffix=''):
...
+        """
+        api_key = request.json.get('api_key')
+        file_id = request.json.get('file_id')
+
+        if not (api_key and file_id):
+            return Response(status=400, json={"message": _("'api_key' or 'file_id' is missing")})
+
+        results = self.get_available_3pm_transcripts(file_id, api_key)
+
+        is_valid = False if isinstance(results, dict) else True
```
Comment (auraz) [inline video_xblock/mixins.py]: Fix it!

File: video_xblock/static/js/studio-edit/studio-edit.js [hunk 27]
```diff
@@ -383,18 +383,76 @@ function StudioEditableXBlock(runtime, element) {
...
+    function validateThreePlayMediaConfig(data) {
+        var message;
+        var options = {
+            type: 'POST',
+            url: runtimeHandlers.validateThreePlayMediaConfig,
+            dataType: 'json',
+            data: JSON.stringify(data)
+        };
+
+        if (!(data.api_key && data.file_id)) {
```
Comment (auraz) [inline video_xblock/static/js/studio-edit/studio-edit.js]: add docstring and comment

File: video_xblock/static/js/studio-edit/studio-edit.js [hunk 28]
```diff
@@ -383,18 +383,76 @@ function StudioEditableXBlock(runtime, element) {
...
+            return {isValid: true};
         }
+
+        return $.ajax(
+            options
+        )
+        .done(function(response) {
+            message = response.message;
```
Comment (auraz) [inline video_xblock/static/js/studio-edit/studio-edit.js]: Why you need message?

File: video_xblock/static/js/studio-edit/studio-edit.js [hunk 29]
```diff
@@ -383,18 +383,76 @@ function StudioEditableXBlock(runtime, element) {
...
+        });
+    }
+
+    /**
+     * Grab 3PlayMedia API configuration data.
+     * @returns (object) 3PlayMedia's: apiKey + fileId
+     */
+    function getThreePlayMediaConfig() {
+        var $apiKey = $('.threeplaymedia-api-key', element).val();
+        var $fileId = $('#xb-field-edit-threeplaymedia_file_id', element).val();
```
Comment (auraz) [inline video_xblock/static/js/studio-edit/studio-edit.js]: I do not think you need $ here.

File: video_xblock/tests/unit/test_utils.py [hunk 30]
```diff
@@ -37,3 +37,22 @@ def test_import_from(self, import_module_mock):
...
+        # Assert:
+        self.assertEqual(reference, expected_ref)
+
+    def test_normalize_transcripts(self):
+        # Arrange
+        test_transcripts = [{'source': 'default'}, {}]
+        # Act
+        normalized_transcripts = normalize_transcripts(test_transcripts)
+        # Assert
+        self.assertEqual(test_transcripts, normalized_transcripts)
```
Comment (z4y4ts) [inline video_xblock/tests/unit/test_utils.py]: So essentially, you are checking, that normalization did nothing with the input data.

I think you need to add few more cases. Or change test name to better reflect your intent.

File: video_xblock/mixins.py (hunk 3 above)
Comment (z4y4ts) [inline video_xblock/mixins.py]: Fields are scoped to a single xblock instance. So no problem here.
